File: leetcode/data_struct/array/0136/1.go
// auther: phoenix
// date: 2023-06-24 21:50:47
package main

import (
	"fmt"
)

/*


 */

func main() {
	singleNumberTest()
}

func singleNumberTest() {
	fmt.Println(singleNumber([]int{2, 2, 1}))       // 1
	fmt.Println(singleNumber([]int{4, 1, 2, 1, 2})) // 4
	fmt.Println(singleNumber([]int{1}))             // 1
	fmt.Println(singleNumber([]int{1, 0, 1}))       // 0

}

func singleNumber(nums []int) int {
	length := len(nums)
	mp := make(map[int]int)

	if length == 1 {
		return nums[0]
	}

	for i := 0; i < length; i++ {
		v := nums[i]
		if _, ok := mp[v]; ok {
			mp[v] = 0
		} else {
			mp[v] = 1
		}
	}

	for k, v := range mp {
		if v == 1 {
			return k
		}
	}
	return -1
}

// https://leetcode.cn/problems/single-number/solution/zhi-chu-xian-yi-ci-de-yuan-su-by-disco-2-q634/
// 异或，两个相同的数异或得到0，将数组所有元素进行异或操作，得到的结果即为只出现一次的元素。
func upSingleNumber1(nums []int) (ans int) {
	// 异或 ^
	for _, v := range nums {
		ans = ans ^ v
	}
	return
}

File: leetcode/data_struct/array/0136/1.ts
// auther: phoenix
// date: 2023-06-24 21:05:54

/*


*/


main()
function main() {
  singleNumberTest()
}

function singleNumberTest(){
  console.log(singleNumber([2,2,1]))      // 1
  console.log(singleNumber([4,1,2,1,2]))  // 4
  console.log(singleNumber([1]))          // 1
  console.log(singleNumber([1,0,1]))      // 0

  
}

function singleNumber(nums: number[]): number {
  let arr:number[] = []
  let length:number = nums.length
  if(nums.length==1) return nums[0]
  for (let i = 0; i < length; i++) {
    const v = nums[i]
    if (arr.includes(v)) {
      let index = arr.findIndex(item=>item === v)
      arr.splice(index,1)
    } else {
      arr.push(v)
    }
  }

  return arr[0]
};



// https://leetcode.cn/problems/single-number/solution/jian-dan-de-jie-fa-by-okisama-qr6b/
function upSingleNumber1(nums: number[]): number {
  return nums.reduce((a, b) => a ^ b)
};
function upSingleNumber2(nums: number[]): number {
  return eval(nums.join('^'))
};

// https://leetcode.cn/problems/single-number/solution/zhao-dao-zhi-chu-xian-yi-ci-de-shu-zi-by-45pc/
function upSingleNumber3(nums: number[]): number {
  /**

异或运算，相异为真，相同为假，所以 a^a = 0 ;0^a = a
因为异或运算 满足交换律 a^b^a = a^a^b = b 所以数组经过异或运算，单独的值就剩下了

      假设 nums = [4,3,3,2,4]
      具体步骤:
          0 ^= 4 -> 4
          4 ^= 3 -> 7
          7 ^= 3 -> 4
          4 ^= 2 -> 6
          6 ^= 4 -> 2
      最后返回 2
   */
  let reduce: number = 0;
  for (let num of nums) {
      reduce = reduce ^ num;
  }
  return reduce;
};
